smtp/init.lua

```diff
@@ -167,15 +181,60 @@ curl_mt = {
                 header = header .. table.concat(opts.headers, '\r\n') .. '\r\n'
             end
 
-            local content_type = opts.content_type or 'text/plain'
-            local charset = opts.charset or 'UTF-8'
-            header = header .. 'Content-Type: ' .. content_type .. '; charset=' .. charset .. '\r\n'
+            local attachments = ''
+            if not opts.attachments or #opts.attachments == 0 then
+                local content_type = opts.content_type or 'text/plain'
+                local charset = opts.charset or 'UTF-8'
+                header = header .. 'Content-Type: ' .. content_type .. '; charset=' .. charset .. '\r\n'
+
+                if opts.subject then
+                    header = header .. 'Subject: ' .. opts.subject .. '\r\n'
+                end
+
+                body = header .. '\r\n' .. body
+            else
+                local MULTIPART_CONTENT_TYPE = 'Content-Type: multipart/mixed; boundary=MULTIPART-MIXED-BOUNDARY;\r\n'
+                local MULTIPART_SEPARATOR = '\r\n--MULTIPART-MIXED-BOUNDARY\r\n'
+                local MULTIPART_END = '\r\n--MULTIPART-MIXED-BOUNDARY--\r\n'
+                for _, attachment in ipairs(opts.attachments) do
+                    local content_type = 'Content-Type: ' ..
+                                         (attachment.content_type or 'text/plain') ..
+                                         '; charset=' ..
+                                         (attachment.charset or  'UTF-8') ..
+                                         ';\r\n'
+                    local content_transfer_encoding = attachment.base64_encode
+                                                      and 'Content-Transfer-Encoding: base64;\r\n'
+                                                      or ''
+                    local content_disposition = 'Content-Disposition: inline; filename=' ..
+                                                attachment.filename ..
+                                                ';\r\n\r\n'
+                    local attachment_body = attachment.base64_encode
+                                            and digest.base64_encode(attachment.body)
+                                            or attachment.body
+                    attachments = attachments ..
+                                  MULTIPART_SEPARATOR ..
+                                  content_type ..
+                                  content_transfer_encoding ..
+                                  content_disposition ..
+                                  attachment_body
+                end
+
+                local content_type = 'Content-Type: ' .. (opts.content_type or 'text/plain') ..
+                                         '; charset=' .. (opts.charset or  'UTF-8') ..';\r\n\r\n'
+
+                if opts.subject then
+                    header = header .. 'Subject: ' .. opts.subject .. '\r\n' .. ''
+                end
```

Comment:
I would perform an attempt to reduce code duplication where possible: either by changing of the code flow or by moving a common functionaliy into functions. Say, a function like `encode_content_type()` may be used as for the message itself as well as for attachments.

Reply:
Mostly resolved in PR #23.